lib/graphql/scalar_type.rb

```diff
@@ -109,7 +109,19 @@ def ensure_two_arg(callable, method_name)
     end
 
     def coerce_non_null_input(value, ctx)
-      @coerce_input_proc.call(value, ctx)
+      @coerce_input_proc.call(raw_coercion_input(value), ctx)
+    end
+
+    def raw_coercion_input(value)
+      if value.is_a?(GraphQL::Language::Nodes::InputObject)
```

Comment:
I'd love to avoid type switches like this by adding a `GraphQL::Language::Nodes::AbstractNode#scalar_conversion_input` method but how you feel about adding methods like this to the AST nodes.